pkg/apis/core/validation/validation.go

```diff
@@ -2521,6 +2521,72 @@ func validatePullPolicy(policy core.PullPolicy, fldPath *field.Path) field.Error
 	return allErrors
 }
 
+func validateEphemeralContainers(ephemeralContainers []core.EphemeralContainer, containers, initContainers []core.Container, volumes map[string]core.VolumeSource, fldPath *field.Path) field.ErrorList {
```

Comment:
You need to validate that pods can't have ephemeral containers with duplicate names to init containers or containers in validatePod

Reply:
I do test for name collisions here in `validateEphemeralContainers()`, which is called from `ValidatePod()`. Is it that I missed an edge case or that you'd like validation moved?

Since containers and init containers cannot be added after creation I didn't expect we'd need to check the container/initcontainer names against the ephemeral containers.

Reply:
ValidatePodTemplateSpec can be called from DaemonSets, deployments, etc which are mutable.

However, as long as this method is always called in those paths, that's ok.

Reply:
Ah, ok. Ephemeral Containers aren't allowed in Pod Templates at all since they shouldn't be a building block of services. I added some unit tests to TestValidateDaemonSet, TestValidateDeployment and TestValidateReplicationController to convince myself that ValidatePodTemplateSpec is being called from these.

I've aded them to this PR since they seem useful to prevent regressions, but I'll leave these in a separate commit to be squashed later in case you tell me to drop them.